server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -118,29 +118,29 @@ public void addDynamic(String name, String prefix, Mapper mapper, DocumentParser
             // otherwise we strip off the first object path of the mapper name, load or create
             // the relevant object mapper, and then recurse down into it, passing the remainder
             // of the mapper name. So for a mapper 'foo.bar.baz', we locate 'foo' and then
-            // call addDynamic on it with the name 'bar.baz'.
+            // call addDynamic on it with the name 'bar.baz', and next call addDynamic on 'bar' with the name 'baz'.
             else {
                 int firstDotIndex = name.indexOf(".");
-                String childName = name.substring(0, firstDotIndex);
-                String fullChildName = prefix == null ? childName : prefix + "." + childName;
-                ObjectMapper.Builder childBuilder = findChild(fullChildName, context);
-                childBuilder.addDynamic(name.substring(firstDotIndex + 1), fullChildName, mapper, context);
-                add(childBuilder);
+                String parentName = name.substring(0, firstDotIndex);
+                String fullParentName = prefix == null ? parentName : prefix + "." + parentName;
+                ObjectMapper.Builder parentBuilder = findParentBuilder(fullParentName, context);
+                parentBuilder.addDynamic(name.substring(firstDotIndex + 1), fullParentName, mapper, context);
+                add(parentBuilder);
```

Comment:
I appreciate that `child` maybe isn't the best name here, but I don't think that `parent` is very good either.  Maybe `immediateChild`? It's a child of the object mapper that `addDynamic` is being called on, so I think calling it `parent` will just confuse things.

Reply:
++ I see what you mean.